### ISSUE TITLE
Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.

### ISSUE THREAD

**Issue by DzonnyDZ** [Microsoft/CodeContracts#20]
When I have somewhat crazy class in my project like this:
```C#
public abstract class EntityDataAccessLayerPart<TDal, TDataContext, TItemBase, TSoftDelete> : EntityDataAccessLayerPart
    where TItemBase : class
    where TSoftDelete : class, TItemBase
    where TDal : EntityDataAccessLayer<TDataContext, TItemBase, TSoftDelete>
    where TDataContext : EntityContextBase
```
When I have code contracts enabled in my project I get following error when compiling
```
1>------ Build started: Project: RFERL.EntityFramework, Configuration: Debug-Fast Any CPU ------
1>  elapsed time: 1672ms
1>C:\Program Files (x86)\Microsoft\Contracts\MsBuild\v12.0\Microsoft.CodeContracts.targets(397,5): warning : Contract reference assembly for project 'RFERL.CoreUtilities' not found. Select 'Build' or 'DoNotBuild' for Contract Reference in project settings.
1>  Reading assembly 'RFERL.EntityFramework' from 'C:\Code\common\RFERL.EntityFramework\obj\Debug-Fast\RFERL.EntityFramework.dll' resulted in errors.
1>  	Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.
1>ccrewrite : error : Rewrite aborted due to metadata errors. Check output window
1>  elapsed time: 789,191ms
1>C:\Program Files (x86)\Microsoft\Contracts\MsBuild\v12.0\Microsoft.CodeContracts.targets(254,5): error MSB3073: The command ""C:\Program Files (x86)\Microsoft\Contracts\Bin\ccrewrite.exe" "@RFERL.EntityFrameworkccrewrite.rsp"" exited with code -1.
========== Build: 0 succeeded, 1 failed, 7 up-to-date, 0 skipped ==========

```
When I comment out part of the code like this
```C#
public abstract class EntityDataAccessLayerPart<TDal, TDataContext, TItemBase, TSoftDelete> : EntityDataAccessLayerPart
    where TItemBase : class
    where TSoftDelete : class, TItemBase
    where TDal : EntityDataAccessLayer//<TDataContext, TItemBase, TSoftDelete>
    where TDataContext : EntityContextBase
```
everything works

**Comment by aarondandy** [comment]
May be related:

- https://social.msdn.microsoft.com/Forums/en-US/acdb74d0-e48d-4f45-89d6-8f845d861559/ccrewriter-crash-unable-to-cast-object-of-type-systemcompilertypeparameter-to-type?forum=codecontracts
- https://github.com/Microsoft/CodeContracts/blob/180c3ae426fae1c31105fc672eeba57c91f76ed0/System.Compiler/StandardVisitor.cs#L1687

**Comment by DzonnyDZ** [comment]
If somebody can advise me which project produces ccrewrite.exe, I can try to fix the issue myself, but I was unable to figure it out myself :-(.

**Comment by FrancescoLogozzo** [comment]
Foxtrot will generate the rewriter.
Then we ILMerge it to generate ccrewrite.exe

**Comment by aarondandy** [comment]
Did you mean to close this?

**Comment by hubuk** [comment]
@DzonnyDZ: can you provide some minimal reproduction project? Type definitions for EntityDataAccessLayerPart, EntityDataAccessLayer<TDataContext,TItemBase,TSoftDelete> and EntityContextBase with all the relevant subtypes may also be sufficient.

The problem is not reproducible on CC taken from Microsoft:master branch at commit 658ad93f26cfd0bb34954e836328efd31aa18842.

**Comment by NikitaEgorov** [comment]
I have some trouble too
```
public interface IIdentityObject<out TIdentity>
{
}
 
public abstract class DefaultDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent>
    where TPersistentDomainObjectBase : class, IIdentityObject<TIdent>, TDomainObjectBase
    where TDomainObjectBase : class
    where TDomainObject : class, TPersistentDomainObjectBase
{
    protected TDomainObject GetNested<TNestedDomainObject>(TDomainObject x) where TNestedDomainObject : TDomainObject
    {
        return default(TDomainObject);
    }
}
 
public abstract class DefaultSecurityDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent> 
    : DefaultDomainBLLBase<TPersistentDomainObjectBase, TDomainObjectBase, TDomainObject, TIdent>
    where TPersistentDomainObjectBase : class, IIdentityObject<TIdent>, TDomainObjectBase
    where TDomainObjectBase : class
    where TDomainObject : class, TPersistentDomainObjectBase
{
}
```

Error	2	The command ""C:\Program Files (x86)\Microsoft\Contracts\Bin\ccrewrite.exe" "@testccrewrite.rsp"" exited with code -1.	test
Error	1	Unable to cast object of type 'System.Compiler.TypeParameter' to type 'System.Compiler.Class'.	c:\users\negorov\documents\visual studio 2013\Projects\test\test\ccrewrite	test

**Comment by hubuk** [comment]
Thanks, problem reproduced. I will try to investigate and fix it.

**Comment by DzonnyDZ** [comment]
So, you no longer need repro project from me. Do you?

**Comment by hubuk** [comment]
Thanks, reproduction project is no longer needed.

**Comment by hubuk** [comment]
Should be fixed in #55.
